ta4j-core/src/main/java/org/ta4j/core/indicators/statistics/VarianceIndicator.java

```diff
@@ -25,53 +25,122 @@
 
 import org.ta4j.core.Indicator;
 import org.ta4j.core.indicators.CachedIndicator;
-import org.ta4j.core.indicators.SMAIndicator;
 import org.ta4j.core.num.Num;
+import org.ta4j.core.num.NumFactory;
 
 /**
- * Variance indicator.
+ * Variance indicator with optimized calculation for sequential access.
+ * Uses Welford's online algorithm for sequential variance calculation.
  */
-public class VarianceIndicator extends CachedIndicator<Num> {
+public final class VarianceIndicator extends CachedIndicator<Num> {
 
     private final Indicator<Num> indicator;
     private final int barCount;
-    private final SMAIndicator sma;
-
-    /**
-     * Constructor.
-     *
-     * @param indicator the indicator
-     * @param barCount  the time frame
-     */
-    public VarianceIndicator(Indicator<Num> indicator, int barCount) {
+    private final boolean isSampleVariance;
+
+    // State for sequential calculation
+    private int previousIndex = -1;
+    private Num previousSum;
+    private Num previousSumOfSquares;
+
+    private VarianceIndicator(final Indicator<Num> indicator, final int barCount, final boolean isSampleVariance) {
         super(indicator);
         this.indicator = indicator;
-        this.barCount = barCount;
-        this.sma = new SMAIndicator(indicator, barCount);
+        this.barCount = Math.max(barCount, 1);
+        this.isSampleVariance = isSampleVariance;
+        this.previousSum = numFactory().zero();
+        this.previousSumOfSquares = numFactory().zero();
+    }
+
+    public static VarianceIndicator ofSample(final Indicator<Num> indicator, final int barCount) {
```

Comment:
I am not convinced this is the implementation pattern we should adopt. I believe I have already voiced my aversion to both private constructors (above) and boolean constructor params (in other PRs and discussions) and this approach would ultimately contain both.

Boolean constructor params are simply less readable than many of the alternatives, and while IDEs that label params can help, we should not rely on them as a crutch for readable code. Brainstorming alternative approaches:

**Use an Enum instead of a boolean**: 
Instead of `new VarianceIndicator(indicator, barCount, true)` 
we'd have `new VarianceIndicator(indicator, barCount, VarianceType.POPULATION)`. 

Someone reading the code in vim or notepad could still reasonably follow along. The enum would be shared among all classes that currently use the boolean flag.

**Separate classes**: This can be implemented in several ways but the naive approach duplicates a lot of code. 

Since the only difference between calculating sample and population is the divisor used, an approach that minimizes code duplication is a "base" `VarianceIndicator` class that contains most of the logic but contains an overridable `getDivisor(int numberOfObservations)` function which it calls when performing the actual variance calculation. 

We then have two classes, `PopulationVarianceIndicator` and `SampleVarianceIndicator` which extend the base class and override the `getDivisor()` function appropriately. If we want to consider the sample variation as the default, `getDivisor()` could return `numberOfObservations - 1`as the default thus eliminating the need for the extra `SampleVarianceIndicator`. 

Other ideas also welcome. Thoughts?

Reply:
I simply agree. I have chosen enum for simplicity. Usually looking for variance and then I have to decide which variant I want to use.